File: src/test/java/im/tox/tox4j/callbacks/FriendStatusMessageCallbackTest.java
package im.tox.tox4j.callbacks;

import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.enums.ToxConnection;
import im.tox.tox4j.exceptions.ToxException;
import im.tox.tox4j.AliceBobTestBase;
import im.tox.tox4j.ToxCore;
import im.tox.tox4j.ToxCoreImpl;
import im.tox.tox4j.ToxOptions;
import im.tox.tox4j.exceptions.ToxNewException;

import static org.junit.Assert.assertEquals;

public class FriendStatusMessageCallbackTest extends AliceBobTestBase {

    @Override
    protected ToxCore newTox(ToxOptions options) throws ToxNewException {
        return new ToxCoreImpl(options);
    }

    @Override
    protected ChatClient newClient() {
        return new Client();
    }


    private static class Client extends ChatClient {

        @Override
        public void friendConnectionStatus(final int friendNumber, @NotNull ToxConnection connection) {
            if (connection != ToxConnection.NONE) {
                debug("is now connected to friend " + friendNumber);
                addTask(new Task() {
                    @Override
                    public void perform(ToxCore tox) throws ToxException {
                        tox.setStatusMessage(("I like " + getFriendName()).getBytes());
                    }
                });
            }
        }

        @Override
        public void friendStatusMessage(int friendNumber, @NotNull byte[] message) {
            debug("friend changed status message to: " + new String(message));
            assertEquals(friendNumber, 0);
            assertEquals("I like " + getName(), new String(message));
            finish();
        }

    }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/callbacks/FriendStatusMessageCallbackTest.scala
package im.tox.tox4j.core.callbacks

import im.tox.tox4j.core.enums.ToxConnection
import im.tox.tox4j.testing.autotest.{AliceBobTest, AliceBobTestBase}

final class FriendStatusMessageCallbackTest extends AliceBobTest {

  override type State = Int
  override def initialState: State = 0

  protected override def newChatClient(name: String, expectedFriendName: String) = new ChatClient(name, expectedFriendName) {

    override def friendConnectionStatus(friendNumber: Int, connectionStatus: ToxConnection)(state: ChatState): ChatState = {
      super.friendConnectionStatus(friendNumber, connectionStatus)(state)
      if (connectionStatus != ToxConnection.NONE) {
        state.addTask { (tox, state) =>
          tox.setStatusMessage(s"I like $expectedFriendName".getBytes)
          state
        }
      } else {
        state
      }
    }

    override def friendStatusMessage(friendNumber: Int, message: Array[Byte])(state: ChatState): ChatState = {
      debug(s"friend changed status message to: ${new String(message)}")
      assert(friendNumber == AliceBobTestBase.FriendNumber)

      state.get match {
        case 0 =>
          assert(message.isEmpty)
          state.set(1)
        case 1 =>
          assert(new String(message) == s"I like $selfName")
          state.finish
      }
    }
  }

}

